### PR TITLE
Use mongodb as optional index for statepoints and job docs

### PR DESCRIPTION
**Pull request by jglaser**
## Description

Use `mongodb` as an optional backend (*via* pymongo). Store state points and job docs in the database instead of creating a directory for every statepoint. Local workspaces are still created by calling `job.open()` explicitly. Jobs are added to the database only with `job.add()`.

To enable the database backend, configure a database host (`myhost`) as described [here](https://docs.signac.io/en/latest/configuration.html#host-configuration) and put these variables into the configuration (`signac.rc`):

```
index_host = myhost
index_db = signac
```

## Motivation and Context
We aim to overcome a scalability issue in signac for large indices (>100,000), when the file system limits the number of subdirectories, i.e. state points, to be created. By decoupling state points from local workspace directories, distributed workers could e.g. create those directories on node-local flash memory as needed for a subset of the entire workspace, while connecting to the central database to store and retrieve job information.

A reference implementation using `mongodb` as a persistent backend on an OLCF OpenShift cluster is underway (actually, figuring out how to allow network access to the cluster took most of the time, rather than implementing these changes :-), which follow ideas by @csadorf.

## Types of Changes

- [ ] Documentation update
- [ ] Bug fix
- [x] New feature
- [x] Breaking change<sup>1</sup>

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:

- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac/blob/master/contributors.yaml).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md#code-style) of this project.
- [ ] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [x] I have updated the API documentation as part of the package doc-strings.
- [ ] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [ ] I have updated the [changelog](https://github.com/glotzerlab/signac/blob/master/changelog.txt) and added all related issue and pull request numbers for future reference (if applicable). See example below.

Example for a changelog entry: `Fix issue with launching rockets to the moon (#101, #212).`

### CODE REVIEW

**Comment by jglaser** [comment]
An obvious bottleneck is now the state point search, which loads all statepoints into the cache first before filtering. We will want to map that onto a `mongodb` query

- [ ] use mongodb query API

File: signac/contrib/project.py [hunk 1]
```diff
@@ -1397,8 +1438,8 @@ def _read_cache(self):
             logger.debug("Read cache in {:.3f} seconds.".format(delta))
             return cache
 
-    def index(self, formats=None, depth=0,
-              skip_errors=False, include_job_document=True):
+    def index_from_workspace(self, formats=None, depth=0,
+                             skip_errors=False, include_job_document=True):
```
Comment (csadorf) [inline signac/contrib/project.py:1442]: I think we should modify the `index()` method such that it calls both `index_from_workspace()` and `index_from_db()` and make the latter two private methods.

**Review by csadorf — CHANGES_REQUESTED** [review]

File: signac/contrib/project.py (hunk 1 above)
Comment (jglaser) [reply signac/contrib/project.py:1442]: That might be a good option. Are you proposing to use the union of the two sets of statepoints then? There could be statepoints that are only local, statepoints that are in the db, and state points that are in both.. need to make the latter unique.

**Comment by b-butler** [comment]
This is handled in a current GSoC project. A MongoDB is one of the many backends for signac synced collections in #364. When the refactoring of data structures is done, adding support for different backends to job statepoints is planned.

**Pull request closed without merging**